### ISSUE TITLE
Table not displaying after a vertical scroll and then a change to different columns and rows.

### ISSUE THREAD

**Issue by professorjoshua** [swimlane/ngx-datatable#647]
[ x ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here

**Current behavior**

Create a drop down list that allows a user to select an object (a different table from the database to display).  When the user selects the object change the column headings, call the API to retrieve the appropriate data.  Up until now everything works.  Even if one object has 10 rows and a second object has 1000 rows.  Now set the vertical scroll to true (to increase the performance).  Switching between objects still works partially.  If you select the object with 1000 items, scroll towards the bottom of the table and then select the object with less rows you will see that the table is not repainted unless you scroll.  You may have to resize the browser window to get a scroll bar.  

**Expected behavior**

When you change the columns and rows (by selecting a different object) then table should be repainted regardless of where you were scrolled to in the previous table.

**Reproduction of the problem**

**What is the motivation / use case for changing the behavior?**

The vertical scroll is needed to increase the performance.  Displaying the table of a 1100 rows of 5 columns took 6 to 10 seconds (depending on the browser) without the vertical scroll.  With the vertical scroll the table was rendered in 1 second.      

**Please tell us about your environment:**

Windows 10 Enterprise, Visual Studio 2015, IIS 

* **Table version:** 0.7.x

"@swimlane/ngx-datatable": "~6.3.0"

* **Angular version:** 2.0.x

Angular 2

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

All browsers

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript

**Comment by adammedford** [comment]
I thought about this an as an interim solution, try adding *ngIf='areColumnsInitialized' and then
```
areColumnsInitialized: boolean = false;

changeColumns(newColumns) {
this.areColumnsInitialized = false;
this.columns = newColumns;
this.areColumnsInitialized = true;
}

```

The idea being that the grid is briefly destroyed and recreated when you change columns. I suspect that is your issue so you wouldn't have to wait on all the rows.

**Comment by professorjoshua** [comment]
Great minds think alike, however, I already tried that to no avail.  In my code I used a Boolean called scrollbarV.  I tried to change the scrollbarV property to true and false.  I also tried the *ngIf based on the Boolean.  The setting to false and true were placed around the assignment of columns.

        <ngx-datatable *ngIf="scrollbarV==true" #myTable style="height:400px;" 
                       [rows]="datarows"
                       [columnMode]="'standard'"
                       [columns]="columnproperties"
                       class='material expandable'
                       [selected]="selected"
                       [selectionType]="'single'"
                       [messages]="messages"
                       (activate)="onActivate($event)"
                       (select)='onSelectItem($event)'
                       [loadingIndicator]="loadingIndicator"
                       [scrollbarV]="scrollbarV"
                       [rowHeight]="50"   
                       id="lookupdata">         
        </ngx-datatable>

**Comment by ipassynk** [comment]
Try to add  setTimeout(() => window.dispatchEvent(new Event('resize'))); after new set is selected.

**Comment by professorjoshua** [comment]
Tried setTimeout(() => window.dispatchEvent(new Event('resize'))); after the new set of data is selected but I still have the same problem.  It appears the ngx-datatable is stuck in the position I scrolled to from the previous dataset.  I even tried to use scrollTop() but have not had any success.

**Comment by enough1987** [comment]
I have the same issue, when I go to another page (the same component is going ) I see white space .

**Comment by arlowhite** [comment]
I've encountered this bug. I thought it was related to Angular change detection, because when I focused a md-input-container, suddenly ngx-datatable displayed. I tried manually calling `markForChanges()` in a timeout, but it didn't work.
I think it may be related to this visibility directive thing:
https://github.com/swimlane/ngx-datatable/blob/908f1216818efe9c8c6322243070eed3cd7479ea/src/directives/visibility.directive.ts

I had another scrolling issue, which I fixed by manually scrolling `datatableBody`, but I think it could work as a hack to fix this issue.

**Setup**
```
  @ViewChild(DatatableComponent)
  dataTable: DatatableComponent;

  private datatableBodyElement: Element;

  ngAfterViewInit(): void {
    this.datatableBodyElement = this.dataTable.element.querySelector('datatable-body');
  }
```
**Scroll**
`    this.datatableBodyElement.scrollTop = offsetY;`

So maybe if you scrolled 1px back and forth it would display the table?

**Comment by professorjoshua** [comment]
@arlowhite Thank you for the suggestion, however, it didn't work for me.  I tried @ipassynk proposed solution too but I'm still stuck.

**Comment by arlowhite** [comment]
@professorjoshua did you scroll after a timeout (setTimeout)? maybe scroll back and forth 1px?

If all else fails, maybe force change detection first with `markForCheck()`, then after 1ms timeout scroll 1px.

**Comment by rodrigoEclipsa** [comment]
The same thing happens to me, I want to use the same table for multiple data and columns...
![image](https://user-images.githubusercontent.com/6819299/27160671-3127b912-514c-11e7-9d8e-77e5c4a6e7e8.png)

**Comment by adammedford** [comment]
I believe this is caused by the same bug as #843. Basically, the rows are rendered in the latter scenario but they retain the scroll offset from before so they are rendered below the container. Scrolling then recalculates those rows positions so they appear in the viewport.

**Comment by rodrigoEclipsa** [comment]
I have had to abandon the use of ngx-datatable, given the many problems and bugs I have encountered. I think that the vitualScroll would have to rethink, since in HTML the problem is not the displacement, if not the deferred creation of the rows. You can move a <table> element with more than 10000 rows and it will be faster than the virtualScroll with 100 rows

**Comment by professorjoshua** [comment]
I have also had to abandon the use of the ngx-datatable.  Sorry.

**Comment by brokenhands** [comment]
After a lot of attempted work-arounds, I've found something that worked for my project. 

For me this bug was manifesting when the user had scrolled beyond the viewport, and we changed the rows array (shifting categories in my case). The new recordset would always appear blank if it was short, or it would be invisible until scrolling for longer recordsets. The code previously reset the rows to being empty while waiting on the service like this:
`
this.table.rows = [];
`
In my case, adding a simple object with a single column property in it was enough:
`    this.table.rows = [{name: ""}];
`

**Comment by jordanburke** [comment]
We ran into the same thing as @brokenhands clearing the results before the next set, and confirming the workaround also works! The only downside is if you have the row count display shown it does count it as a single row in the interim.

**Comment by markanthonyg** [comment]
Finally found a ridiculous hack for this:

```
http.get(...).subscribe(res => {
     /// set rows
     setTimeout(function() { $('datatable-body').scrollTop(0); }, 1);
     setTimeout(function() { $('datatable-body').scrollTop(1); }, 1);
}
```

Works in chrome + ie.
Using angular 5.0

**Comment by rodrigoEclipsa** [comment]
a solution with a timer .. better step ....

**Comment by LeonardoWiest** [comment]
It's working for me now, after adding this source code:

`setTimeout(() => {
                        this.myGrid..element.getElementsByTagName('datatable-body')[0].scrollTop++;
                    }, 1);`

**Comment by elcsiga** [comment]
It seems that if we remove some rows and scroll bar disappears, the data table element does not reset the scrollPos values itself. So I ended up adding something like this in my datatable component which uses ngx-data-table:

    @Input() set data(value: any[]) {
        this.rows = value;
 
        setTimeout(() => {
            // here we reset scrollPos values on each data change if there is no scroller present.
            if (!this.dataTable.bodyComponent.scroller) {
                this.dataTable.bodyComponent.onBodyScroll({
                    scrollYPos: 0,
                    scrollXPos: 0
                });
            }
        }, 0);
    }
